File: tictactoe_app/src/main/java/com/app/tictactoe/controller/AuthController.java
package com.app.tictactoe.controller;

import com.app.tictactoe.model.User;
import com.app.tictactoe.service.UserService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.*;

@Controller
public class AuthController {

    @Autowired
    private UserService userService;

    @GetMapping("/")
    public String dashboard(Model model) {
        User user = getCurrentUser();
        model.addAttribute("user", user);
        return "dashboard";
    }

    @GetMapping("/register")
    public String showRegistrationForm(Model model) {
        model.addAttribute("user", new User());
        return "register";
    }

    @PostMapping("/register")
    public String registerUser(@ModelAttribute User user, Model model) {
        try {
            userService.registerUser(user.getUsername(), user.getPassword(), user.getEmail());
            return "redirect:/login";
        } catch (IllegalArgumentException e) {
            model.addAttribute("error", e.getMessage());
            return "register";
        }
    }

    @GetMapping("/login")
    public String login(@RequestParam(value = "error", required = false) String error, Model model) {
        if (error != null) {
            model.addAttribute("error", "Invalid username or password.");
        }
        return "login";
    }

    @PostMapping("/logout")
    public String logout() {
        return "redirect:/login?logout";
    }

    @GetMapping("/profile")
    public String showProfilePage(Model model) {
        User user = getCurrentUser();
        model.addAttribute("user", user);
        return "profile";
    }

    @PostMapping("/profile")
    public String updateUsername(@RequestParam String username, Model model) {
        User user = getCurrentUser();
        try {
            userService.updateUsername(user.getId(), username);
            updateAuthentication(user, username);
            model.addAttribute("success", "Username updated successfully");
        } catch (IllegalArgumentException e) {
            model.addAttribute("error", e.getMessage());
        }
        model.addAttribute("user", userService.findById(user.getId()));
        return "profile";
    }

    private User getCurrentUser() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        String username = authentication.getName();
        return userService.findByUsername(username);
    }

    private void updateAuthentication(User user, String newUsername) {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        UsernamePasswordAuthenticationToken newAuth = new UsernamePasswordAuthenticationToken(
                newUsername, authentication.getCredentials(), authentication.getAuthorities());
        SecurityContextHolder.getContext().setAuthentication(newAuth);
    }
}

File: tictactoe_app/src/main/kotlin/com/app/tictactoe/controller/AuthController.kt
package com.app.tictactoe.controller

import com.app.tictactoe.model.User
import com.app.tictactoe.service.UserService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.web.bind.annotation.*

@Controller
class AuthController @Autowired constructor(
    private val userService: UserService
) {

    @GetMapping("/")
    fun dashboard(model: Model): String {
        val user = getCurrentUser()
        model.addAttribute("user", user)
        return "dashboard"
    }

    @GetMapping("/register")
    fun showRegistrationForm(model: Model): String {
        model.addAttribute("user", User())
        return "register"
    }

    @PostMapping("/register")
    fun registerUser(@ModelAttribute user: User, model: Model): String {
        return try {
            userService.registerUser(user.username, user.password, user.email)
            "redirect:/login"
        } catch (e: IllegalArgumentException) {
            model.addAttribute("error", e.message)
            "register"
        }
    }

    @GetMapping("/login")
    fun login(
        @RequestParam(value = "error", required = false) error: String?,
        model: Model
    ): String {
        if (error != null) {
            model.addAttribute("error", "Invalid username or password.")
        }
        return "login"
    }

    @PostMapping("/logout")
    fun logout(): String {
        return "redirect:/login?logout"
    }

    @GetMapping("/profile")
    fun showProfilePage(model: Model): String {
        val user = getCurrentUser()
        model.addAttribute("user", user)
        return "profile"
    }

    @PostMapping("/profile")
    fun updateUsername(
        @RequestParam username: String,
        model: Model
    ): String {
        val user = getCurrentUser()
        return try {
            userService.updateUsername(user.id, username)
            updateAuthentication(username)
            model.addAttribute("success", "Username updated successfully")
            model.addAttribute("user", userService.findById(user.id))
            "profile"
        } catch (e: IllegalArgumentException) {
            model.addAttribute("error", e.message)
            model.addAttribute("user", userService.findById(user.id))
            "profile"
        }
    }

    private fun getCurrentUser(): User {
        val authentication: Authentication = SecurityContextHolder.getContext().authentication
        val username = authentication.name
        return userService.findByUsername(username)
    }

    private fun updateAuthentication(newUsername: String) {
        val authentication: Authentication = SecurityContextHolder.getContext().authentication
        val newAuth = UsernamePasswordAuthenticationToken(
            newUsername, authentication.credentials, authentication.authorities
        )
        SecurityContextHolder.getContext().setAuthentication(newAuth)
    }
}
